lib/puppet/provider/cs_stonith/pcs.rb

```diff
@@ -0,0 +1,209 @@
+  # of actually doing the work.
+  # The existing_resource is there because pcs does not have a single command that
+  # updates or create a resource, so we flag the resources with that parameter
+  def create
+    @property_hash = {
+      :name              => @resource[:name],
+      :ensure            => :present,
+      :primitive_type    => @resource[:primitive_type],
+      :existing_resource => :false
+    }
+    @property_hash[:device_options] = @resource[:device_options] if ! @resource[:device_options].nil?
+    @property_hash[:operations] = @resource[:operations] if ! @resource[:operations].nil?
+  end
+
+  # Unlike create we actually immediately delete the item.
+  def destroy
+    debug('Removing STONITH fence')
+    Puppet::Provider::Pacemaker::run_pcs_command([command(:pcs), 'stonith', 'delete', @property_hash[:name]])
+    @property_hash.clear
+  end
+
+  # Getters that obtains the device_options and operations defined in our primitive
+  # that have been populated by prefetch or instances (depends on if your using
+  # puppet resource or not).
+  def primitive_type
+    @property_hash[:primitive_type]
+  end
+
+  def device_options
+    @property_hash[:device_options]
+  end
+
+  def operations
+    @property_hash[:operations]
+  end
+
+  # Our setters for device_options and operations.  Setters are used when the
+  # resource already exists so we just update the current value in the
+  # property_hash and doing this marks it to be flushed.
+  def fence_type=(should)
+    @property_hash[:fence_type] = should
+  end
+
+  def device_options=(should)
+    @property_hash[:device_options] = should
+  end
+
+  def operations=(should)
+    @property_hash[:operations] = should
+  end
+
+  # Flush is triggered on anything that has been detected as being
+  # modified in the property_hash.
+  # It calls several pcs commands to make the resource look like the
+  # params.
+  def flush
+    unless @property_hash.empty?
+      # The primitive_type variable is used to check if one of the class,
+      # provider or type has changed
+      primitive_type = "#{@property_hash[:primitive_type]}"
+
+      unless @property_hash[:device_options].empty?
+        device_options = []
+        @property_hash[:device_options].each_pair do |k,v|
+          device_options << "#{k}=#{v}"
+        end
+      end
+
+      unless @property_hash[:operations].empty?
+        operations = []
+        @property_hash[:operations].each do |o|
+          op_name = o[0]
+          operations << "op"
+          if op_name.include? ':'
+            items = op_name.split(':')
+            operations << items[0]
+            operations << "role=#{items[1]}"
+          else
+            operations << op_name
+          end
+          o[1].each_pair do |k,v|
+            operations << "#{k}=#{v}"
+          end
+        end
+      end
+
+      # We destroy the resource if it's primitive type has changed
+      unless @property_hash[:existing_resource] == :false
```

Comment:
The `Puppet::Transaction::ResourceHarness#sync_if_needed` method seems to not handle false values correctly; instead of doing a `param.should.nil?` it does a boolean check on `param.should` (https://github.com/puppetlabs/puppet/blob/master/lib/puppet/transaction/resource_harness.rb#L128) which means it'll treat a param with a value of false:falseClass as unset. So, it's kinda broken. I think.